### ISSUE TITLE
Fat-line clipping algorithm sometimes finds intersections and wrong parameters.

### ISSUE THREAD

**Issue by lehni** [paperjs/paper.js#795]
This is probably the reason why example 23 fails in #784:
[Demo Sketch](http://sketch.paperjs.org/#S/lVTfa9swEP5XRF7sUE/ot2SXPXUb9GEw6GNaiusoiakrD9lJByX/+05y0thtOjrbBHT33d13353yMnPlk50Vs5tH21ebWTar2mU472r7jO+fyt7Xf9BX5Owz+hkP6YIZLBjVJBecCHiJzhCJ3znPF6qF4VgQo43RTKocbEoTiZUUVOZMa6X53fzy1t26XelRtfU7Sw81r8IhVZCYcCkAaATNc0jBBAeTMlITajTnlGUowogQTDElNRMKYJJjzTgVDOIl0QEGGTDTJgejkYRQIMmVABPRWgptJNdGfxIWaL+SZlPSEC8Z8OFSEa1yA2wYxUZLTYmQRFJjAulPwkJv0JUGJhKKg4lxhhURBiDSMNBRHmD/Ukrlmg1ar7au6uvWocrbsre/yn6TxjYyVLVN6+fo5dYheLztt97FziLoYA5PZ9dP1vVdgRYxFB8MNEOTM7vLRkG9bx/tVahRDKXeOW+qsqndukCrsuns4N0H2vszzK9qXzU2rV1vT6TDUH4DWxhJdOBwuBx8Veu6trG4addpElpC198KlKCLGIHr5YnOBUoysHq4JJBkAB3zHYxvwS34J8BgmF+e1RIfuI8krayLlUbBI3l8uay3IDfB9CNFYw8jy3+KexQu3MC3mwFjTbxdJsetDzh2BgfLmTw0W5ucbnVsp7Nxbh2ExBJ4bfvrsSONGUPUZEY/2q1bniQ9onFj3RqmB7pPHUUsPMWuWv+9rDbpcXmGfclQ/boyk5JjWkPlUGZIHMHvV+9DDx4ziUKHH/ivfQDoYxxxNysWd/u/)

Output:
```
Found 2 intersections:
Intersection 0:
Path ID: 1, parameter: 0, point: { x: 628.03547, y: 243.03686 }
Path ID: 3, parameter: 1, point: { x: 628.03547, y: 243.03686 }
Intersection 1:
Path ID: 1, parameter: 5.668521207979893e-13, point: { x: 628.03547, y: 243.03686 }
Path ID: 3, parameter: 0.9434552849029103, point: { x: 628.035, y: 242.8367 }
```

Notice how in the second intersection, the first location on Path 1 is basically the same as in the first intersection, while the location on Path 3 varies. And there is definitely no intersection at  parameter 0.9434552849029103 on Path 3!

@iconexperience let me know if you feel like debugging this one, since you have some experience with the fat-line clipping code. This may well be the source of some of the other failing edge cases too.

Code:

```js
view._matrix = new Matrix([28.421709430404007, 0, 0, 28.421709430404007, -17483.40878872569, -6705.654159277673]);

var curve1 = new Curve(628.0354673841999, 243.03685701873312, 628.0044262657246, 253.72314203550712, 419.2789142850017, 364.27077547853787, 419.2789142850017, 364.27077547853787);
var curve2 = new Curve(619.5200435607698, 221.87571045051882, 619.5200435607698, 221.87571045051882, 628.0657700157852, 232.60480515822775, 628.0354673841999, 243.03685701876972);

function createPath(curve, color) {
    return new Path({
        segments: [curve.segment1, curve.segment2],
        strokeColor: color,
        strokeScaling: false
    });
}

function createCircle(inter) {
    var path = inter.path;
    console.log('Path ID: ' + path.id
        + ', parameter: ' + inter.parameter
        + ', point: ' + inter.point);
    return new Path.Circle({
        center: inter.point,
        radius: 0.1,
        strokeColor: path.strokeColor,
        strokeScaling: false
    });
}

var path1 = createPath(curve1, 'red');
var path2 = createPath(curve2, 'blue');

var intersections = path1.getIntersections(path2);
console.log('Found ' + intersections.length + ' intersections:');
intersections.forEach(function(inter, i) {
    console.log('Intersection ' + i + ':');
    createCircle(inter);
    createCircle(inter.intersection);
});
```

**Comment by iconexperience** [comment]
@hkrish 
Hello Hari, I think we could use your help here.

I was always thinking that it may be worth trying to switch coordinates for each recursion in fatline clipping. I implemented this and it works surprisingly well, as it solves not only this issue but four other of the 34 remaining glitches in my test case (without creating new issues).

What do you think, is this really a good idea or are my results only good by sheer luck?

Here is a quick implementation that I am using in the `addCurveIntersections()` function:

```javascript
  function addCurveIntersections(v1, v2, c1, c2, locations, param,
			tMin, tMax, uMin, uMax, oldTDiff, reverse, recursion) {
    if (++recursion >= 32)
			return;
    var isTransposed = recursion%2 == 1; // transpose x and y for every second recursion
    var transpose = function(v) {
      return [v[1], v[0], v[3], v[2], v[5], v[4], v[7], v[6]];
    }
		
    if (isTransposed) v1 = transpose(v1), v2 = transpose(v2);
		var q0x = v2[0], q0y = v2[1], q3x = v2[6], q3y = v2[7],
			epsilon = 0.000001,
			getSignedDistance = Line.getSignedDistance,
			d1 = getSignedDistance(q0x, q0y, q3x, q3y, v2[2], v2[3]) || 0,
			d2 = getSignedDistance(q0x, q0y, q3x, q3y, v2[4], v2[5]) || 0,
			factor = d1 * d2 > 0 ? 3 / 4 : 4 / 9,
			dMin = factor * Math.min(0, d1, d2),
			dMax = factor * Math.max(0, d1, d2),
			tMinNew, tMaxNew,
			tDiff;
    var dp0 = getSignedDistance(q0x, q0y, q3x, q3y, v1[0], v1[1]),
      dp1 = getSignedDistance(q0x, q0y, q3x, q3y, v1[2], v1[3]),
      dp2 = getSignedDistance(q0x, q0y, q3x, q3y, v1[4], v1[5]),
      dp3 = getSignedDistance(q0x, q0y, q3x, q3y, v1[6], v1[7]),
      hull = getConvexHull(dp0, dp1, dp2, dp3),
      top = hull[0],
      bottom = hull[1],
      tMinClip, tMaxClip;
    if ((tMinClip = clipConvexHull(top, bottom, dMin, dMax)) == null ||
      (tMaxClip = clipConvexHull(top.reverse(), bottom.reverse(),
        dMin, dMax)) == null)
      return;
    if (isTransposed) v1 = transpose(v1), v2 = transpose(v2);  // undo transposition
    v1 = Curve.getPart(v1, tMinClip, tMaxClip);
    tDiff = tMaxClip - tMinClip;
    tMinNew = tMax * tMinClip + tMin * (1 - tMinClip);
    tMaxNew = tMax * tMaxClip + tMin * (1 - tMaxClip);
		if (oldTDiff > 0.5 && tDiff > 0.5) {
			if (tMaxNew - tMinNew > uMax - uMin) {
				var parts = Curve.subdivide(v1, 0.5),
					t = tMinNew + (tMaxNew - tMinNew) / 2;
				addCurveIntersections(
					v2, parts[0], c2, c1, locations, param,
					uMin, uMax, tMinNew, t, tDiff, !reverse, recursion);
				addCurveIntersections(
					v2, parts[1], c2, c1, locations, param,
					uMin, uMax, t, tMaxNew, tDiff, !reverse, recursion);
			} else {
				var parts = Curve.subdivide(v2, 0.5),
					t = uMin + (uMax - uMin) / 2;
				addCurveIntersections(
					parts[0], v1, c2, c1, locations, param,
					uMin, t, tMinNew, tMaxNew, tDiff, !reverse, recursion);
				addCurveIntersections(
					parts[1], v1, c2, c1, locations, param,
					t, uMax, tMinNew, tMaxNew, tDiff, !reverse, recursion);
			}
		} else if (Math.max(uMax - uMin, tMaxNew - tMinNew) < epsilon) {
			var t1 = tMinNew + (tMaxNew - tMinNew) / 2,
				t2 = uMin + (uMax - uMin) / 2;
			addLocation(locations, param,
				reverse ? v2 : v1, reverse ? c2 : c1, reverse ? t2 : t1, null,
				reverse ? v1 : v2, reverse ? c1 : c2, reverse ? t1 : t2, null);
		} else if (tDiff > 1e-12) {
			addCurveIntersections(v2, v1, c2, c1, locations, param,
					uMin, uMax, tMinNew, tMaxNew, tDiff, !reverse, recursion);
		}
	}
```

**Comment by lehni** [comment]
@iconexperience that's a great find! I'd suggest since it's not breaking anything at your end and solves a few glitches, why don't you create a PR for it, and I continue the testing with the couple of tests I have running at my end (e.g. https://github.com/paperjs/boolean-test). If we find that it breaks something we can always revert. And if @hkrish thinks it's a bad idea, or has a better suggestion, we can switch to that then. Sounds good?

**Comment by lehni** [comment]
I wonder though how you can explain that this solves it? And I think I have an idea for how to implement with this without the `transpose()` method, so you probably don't need to spend the time on the PR.

**Comment by iconexperience** [comment]
With "great find" you mean <a href="https://www.youtube.com/watch?v=UMRo5XCKddQ">this</a>?

Yes, sounds good. You should also note that I removed <a href="https://github.com/paperjs/paper.js/blob/develop/src/path/Curve.js#L1244">Curve.js lines 1244 to 1250</a> as I thought this would not be necessary anymore. We should keep an eye on this and think deeply

**Comment by lehni** [comment]
Hehe, yes pretty much like that! Now that's interesting... What appears to have fixed issue isn't the *transposing*, it's the  <a href="https://github.com/paperjs/paper.js/blob/develop/src/path/Curve.js#L1244">Curve.js lines 1244 to 1250</a> that you removed! Even without *transposing* the coordinates, I still get the correct result in example 23 without those lines. I'll do some more tests and compare the results of the two approaches in your test suite now.

**Comment by lehni** [comment]
So the *transposing* isn't needed, all it took was the removal of that specific block of code. With it gone, I get this result:

    new glitches = []
    solved glitches = [473, 1412, 2217, 4692, 5606, 5856]

**Comment by iconexperience** [comment]
I can confirm that the transposition does not make a difference if we remove that block. I thought that I had tested that, but obviously not.
Great, next one please :smile:

**Comment by lehni** [comment]
Man I *should* really be working on other things, but this is weirdly addicting...

**Comment by hkrish** [comment]
Interesting! 
Those lines were there, since we have encountered issues before because, one of the fatlines were converging early. Unfortunately I don't remember which was the failing case. So please keep an eye out, if we come across something like that in the future.

The wrong parameters are returned because, when only one of the ranges are  *nearly* converged, the clipping gets very unstable, and we have already lost significant bits in the clipping process. The averaged out coordinates might give imprecise results. 

We can also do this:  If one of the ranges is reduced to 0, evaluate that point in the corresponding curve, and find the nearest parameter on the second curve using the point. This will give better precision than merely averaging, limited only by the precision of the point projection algorithm. Again need to test, this is just an idea.

So, something like
@ https://github.com/paperjs/paper.js/blob/develop/src/path/Curve.js#L1294

```javascript
    else if (Math.max(uMax - uMin, tMaxNew - tMinNew) < tolerance || tDiff == 0) {
         if (tDiff == 0) {    // Special case to handle early convergence
             var pnt = (reverse ? curve2 : curve1).getPointAt(tMinNew)
             var u = (reverse ? curve1 : curve2).getParameterOf(pnt)   // Ideally this should be the parameter of the nearest/projected point,
            ... // Add  {tMinNew, u} to the intersection list
        } else {
            ... // Take average values of t and u
```


I had some code (in the original form, before integrating it with paperjs) lying around somewhere. I will try to find it; if I had some pointers for that.

Glad it's working now.
Cheers.

**Comment by lehni** [comment]
@hkrish I was thinking about doing something like this when first encountering the issue today! Very good to hear your thoughts and explanation on this. I'll test this approach right now and see if its solves more of our edge cases.

**Comment by hkrish** [comment]
Thanks @lehni .
Please note that, getParamterOf might return void since the query-point might be very close to but not on the second curve; so probably we need to project the point with some small tolerance instead.

**Comment by lehni** [comment]
@hkrish I'm using `Curve#getNearestLocation()` for testing now... But I'm wondering, should the comparison be `tDiff == 0`, or more like `tDiff < 1e-12` ?

**Comment by lehni** [comment]
Oh wait, I've made a mistake in my implementation of this... More analysis is needed.

**Comment by hkrish** [comment]
Since we are explicitly setting tDiff = 0  earlier, we just need to catch that.
So tdiff == 0 is good.

**Comment by lehni** [comment]
@hkrish I've found a strange occurrence  where tDiff ends up being negative:

[Demo Sketch](http://sketch.paperjs.org//#S/lVTbattAEP2VxS+Sibrs/aLQJ7eFPBQKeYxDUOSNIyKvykp2CiH/3tmVHcuJU6iQDDNzZs7ZmVm/zHy1cbNydv3khvpxVszqbhXtXeOe8d2mGkLzB31F3j2jn8nIDdFYWy6VskJxqgpE0vvR/4XJaGIqBNdWE0ZYdHKircJMMCq1VlKa+eXSL/2uCqjehp2je75FNHJOBRaWcqkZo4obIwvEjMTGEMOopVYApyhQwkXLMKnhYxGmMOECaLWRTFjLRxjXShlLiSaSHmBUc6sVkwQI1L6aNpRYbQzn0ZVItVBUCiukFFpG3W+q2T9V2zOi9yyGUGmktIAaQZYIzSTlWnBuI4jjqI0oI6EUjSCDOUgncCTLuUp6JWYUShHGJbMaGsKswJoTqayhQiqQMfb5Yevroek8qoOrBverGh7zdIIC1V3bhTl6WXoET3DDNvh0qATau+PTu/XG+aEv0U1KxXsHiDux2W0xSRpC9+QWkaMcqT4Er+uqbfy6RA9V27sx+hplv55RvmhC3bq88YM7io7z+A1qYRopgKNxOcbqzvdd63DbrfMsHgldfStRhi5SBm5WRzkXKCvAG+ByQJERNNa7e/O+R3cAmCKTY355tpl4L37S09r5RDVJnvQnVKtmC/0msKqf9TSdYuL5z/YeWhfv3/vdgMFmwa2yw8pHHDuDg+uU3bdblx3vdDpP79LkekhJFHjthqtpIE8VY9bJlH50W7869vSAxq3za5gfNP40UCbiU+xDF75X9WN+WJ9xYwrUvC3NCeVU1sgcacbCCfxx+T6N4KmS1Oj4A/+y9wB9SjPuZ+XN7etf)

With these two curves, at iteration 20 I end up with a `tMaxClip` that's smaller than the `tMinClip`. Is this supposed to happen? It seems to be the source of an issue in @iconexperience's test suite.

**Comment by lehni** [comment]
BTW: On the `boolean-fix` branch, the two locations get merged to one, but the strange behavior is still there, and somehow it is the origin of a glitch.

**Comment by hkrish** [comment]
Hmmm. That should not happen.

I would try to plot the convex hull being clipped, and the dMin and dMax values, so that we can visualise.

**Comment by lehni** [comment]
@hkrish I don't really understand the convex hull code. @iconexperience, do you understand it?

**Comment by hkrish** [comment]
You can just print out the values of dq0, dq1, dq2 … that is used to create the convex hull and just make a simple drawing. 
Probably we will know why the issue happens there.

**Comment by lehni** [comment]
they are all very small:

    2.108083592442785e-13, 1.199305766274425e-13, -1.1865074770754798e-13, -5.805256553611439e-13

**Comment by lehni** [comment]
I've reintroduced an old version of this code and the problem disappears. It looks like it's caused by later refactoring. Will track it down

**Comment by lehni** [comment]
Holy smokes. The old convex-hull clipping code reduces the failing cases to half. @iconexperience, we have 16 more to go!

**Comment by hkrish** [comment]
Ok. I missed what was changed in the convex hull code. I will check up with you guys later.

**Comment by lehni** [comment]
OK that was too fast... I didn't check against the full set of tests, only the ones that failed previously. While it fixes some of those, it introduces tons of new ones. But it seems that something is wrong in the current implementation as well. It's way above my head to fix though, unfortunately.

**Comment by lehni** [comment]
OK I rushed to conclusions. The version of `clipConvexHull()` I reverted to had a different signature and hence was called wrongly... Now I reverted back to the version right before before we switched to @iconexperience's proposal in 477527f8ec8c348075921ffad827c1ef368c1ee2, and while this solved one issue (5943), it introduced more new ones (814, 2511, 2632, 3897, 7649). Oh dear.

**Comment by iconexperience** [comment]
I though that I had understood the convex hull code, as i implemented the version that appeared to be more intuitive. It is really not that difficult and I will take a closer look again to understand how there can be a negative tDiff, as I always assumed that this cannot happen. Maybe there really is something that needs to be fixed and we can get rid of more glitches...

**Comment by iconexperience** [comment]
@hkrish @lehni
Since `getSignedDistance()` is crucial for the fat line clipping I modified this a little bit, trying to achieve higher precision for points that are very close to the line.

Here is my first modified version:
```javascript
getSignedDistance: function(px, py, vx, vy, x, y, asVector) {
  if (!asVector) {
    vx -= px;
    vy -= py;
  }
  return Numerical.isZero(vx)
      ? vy >= 0 ? px - x : x - px
      : Numerical.isZero(vy)
        ? vx >= 0 ? y - py : py - y
        : (vx * y - vx * py - vy * x + vy * px) / Math.sqrt(vx * vx + vy * vy);
}
```

This needs two additional multiplications, but it solves two more glitches! @hari Maybe you have an idea on how to further imrove precision.

**Comment by iconexperience** [comment]
Well, my modification actually solves 9 glitches, but creates 7 new ones. This could be coincidence, but it shows that `getSignedDistance()` is a hotspot.

**Comment by lehni** [comment]
Interesting! And I was wrong about [5943], this wasn't solved by the older version of `clipConvexHull()`, it was solved by this: c8132584b966fdc95c1773561de4cb7c9c8cc718

**Comment by lehni** [comment]
If I remove the special handling of `vx` or `vy` being close to `0`, I get this:

    new glitches = [796, 7620]
    solved glitches = [6398, 6861]

Hot spot indeed...

**Comment by iconexperience** [comment]
Here is another version which solves two more issues, down to 24:

```javascript
getSignedDistance: function(px, py, vx, vy, x, y, asVector) {
  if (!asVector) {
    vx -= px;
    vy -= py;
  }
  if (Numerical.isZero(vx)) {
      return vy >= 0 ? px - x : x - px;
  } else if  (Numerical.isZero(vy)) {
      return vx >= 0 ? y - py : py - y;
  } else {
      var a = Math.abs(vx) < 1 ? vx * (y - py) : vx * y - vx * py;
      var b = Math.abs(vy) < 1 ? vy * (x - px) : vy * x - vy * px;
      return (a - b) / Math.sqrt(vx * vx + vy * vy);
  }
}
```

**Comment by iconexperience** [comment]
Now this is getting a bit bizarre. Just to check all possible parameters I set <a href="https://github.com/paperjs/paper.js/blob/boolean-fix/src/path/Curve.js#L1448">factor</a> in fat line clipping to 1 and one glich got solved. Down to 23. But why???

**Comment by iconexperience** [comment]
Hoho, setting factor to `.7501` instead of `3/4` solved the glitch, too.

`factor = d1 * d2 > 0 ? .7501 : 4 / 9`

**Comment by lehni** [comment]
We should not forget that a glitch disappearing doesn't necessarily mean it's solved. We should look at the resulting paths, and consider that glitches might simply get masked by something else happening. The best to do is probably still to find out if a given glitch is caused by wrong intersections, and only then try and fix it with improving the intersection code.

**Comment by iconexperience** [comment]
I agree, but I am just collection factors that play a role. And in the 3/4 case, well the threshold was probably a little bit too low :)

**Comment by lehni** [comment]
What I would like to understand first is how we end up getting a value for `tMaxClip` that is smaller than `tMinClip`. This shouldn't happen, regardless of what `getSignedDistance()` returns, right?

**Comment by iconexperience** [comment]
I just found that setting `epsilon=1e-8` lowers the number of reported glitches to 19. Further lowering epsilon seems to introduce more glitches, so at least for our tests `1e-8`seems to be the optimal value. 

Regarding the negative `tDiff` you are correct, we should find the reason for this as something obviously is not working as it should be.

**Comment by lehni** [comment]
Oh interesting! You mean the epsilon value inside `addCurveIntersections()` that currently is set to `1e-6`?

**Comment by iconexperience** [comment]
Yes, that one.

**Comment by lehni** [comment]
What happens at `1e-7`?

**Comment by iconexperience** [comment]
`1e-7` results in 21 glitches.

And just for the record, raising the maximum number of recusions does not change anything.

**Comment by iconexperience** [comment]
Here are some more findings:

1. I was able to simplify `getConvexHull()` quite a bit. Very nice, results seem to be the same and the new implementation makes a lot of sense to me. Here is the code, explanation will follow later:

```javascript
function getConvexHull(dq0, dq1, dq2, dq3) {
  var p0 = [ 0, dq0 ],
    p1 = [ 1 / 3, dq1 ],
    p2 = [ 2 / 3, dq2 ],
    p3 = [ 1, dq3 ],
    getSignedDistance = Line.getSignedDistance,
    abs = Math.abs,
    dist1 = getSignedDistance(0, dq0, 1, dq3, 1 / 3, dq1),
    dist2 = getSignedDistance(0, dq0, 1, dq3, 2 / 3, dq2),
    flip = false,
    hull;
  if (dist1 * dist2 < 0) {
    hull = [[p0, p1, p3], [p0, p2, p3]];
    flip = dist1 < 0;
  } else {
    var distRatio = dist1 / dist2;
    if (distRatio >= 2) {
       hull = [[p0, p1, p3], [p0, p3]]
    } else if (distRatio <= .5) {
       hull = [[p0, p2, p3], [p0, p3]]
    } else {
       hull = [[p0, p1, p2, p3], [p0, p3]];
    }
    flip = dist1 ? dist1 < 0 : dist2 < 0;
  }
  return flip ? hull.reverse() : hull;
}
```

2. The negative tDiff seems to be simply a numerical precision issue. Probably not much we can do here, but I also do not think this really is a problem, as values are around `1e-16` when they are negative.

4. Removing the <a href="https://github.com/paperjs/paper.js/blob/boolean-fix/src/path/Curve.js#L1513">final condition</a> in `addCurvesIntersections()` reduces the number of glitches by one, but causes one case in my test kit to take twenty seconds for fat line clipping. I will try to extract this case and investigate further.

I have to do some other things this week, but whenever I find some time I will continue on this issue.

@lehni Should be open a new issue "Improve fatline clipping"?

**Comment by lehni** [comment]
@iconexperience very interesting! The code looks great, I'm testing it right now. And yes, a new issue seems to be in order. Could you open it? Regarding your points:

1. Are you sure? I added this statement after the calculation of `tDiff`: 

    if (tDiff < 0)
        console.error(tDiff);

And with the above example sketch in https://github.com/paperjs/paper.js/issues/795#issuecomment-145607391, I get this logged: `-0.13331066588893975` With the previous implementation of `getConvexHull()`, it was `-0.2101688247718` so behavior has changed!

**Comment by iconexperience** [comment]
My implementation of `getConvexHull()`above was wrong because I mixed vertical and orthogonal distances. Below is the fixed implementation.

This implementation uses the vertical distances of p1 and p2 to the line between p0 and p3, while our current implementation uses the orthogonal distances calculated by `getSignedDistance()`. The point p2 (at 2/3) is within a triangle defined by p0, p1 and p3 if its dist2 (the distance between p2 and the line between p0 and p3) is less or equal half the distance of dist1 and if dist1 and dist2 have the same sign (green part of line at 2/3). The same can be done with p1 for the triangle between p0, p2, p3.

![image](https://cloud.githubusercontent.com/assets/2122773/10306028/62e8601e-6c24-11e5-8ffa-6f78ab5485b8.png)

This implementation should be slightly faster than our current one.

```javascript
function getConvexHull(dq0, dq1, dq2, dq3) {
  var p0 = [ 0, dq0 ],
    p1 = [ 1 / 3, dq1 ],
    p2 = [ 2 / 3, dq2 ],
    p3 = [ 1, dq3 ],
    getSignedDistance = Line.getSignedDistance,
    dist1 = dq1 - (dq0 + (dq3 - dq0) / 3),
    dist2 = dq2 - (dq0 + 2 * (dq3 - dq0) / 3),
    flip = false,
    hull;
  if (dist1 * dist2 < 0) {
    hull = [[p0, p1, p3], [p0, p2, p3]];
    flip = dist1 < 0;
  } else {
    var distRatio = dist1 / dist2;
    if (distRatio >= 2) {
       hull = [[p0, p1, p3], [p0, p3]]
    } else if (distRatio <= .5) {
       hull = [[p0, p2, p3], [p0, p3]]
    } else {
       hull = [[p0, p1, p2, p3], [p0, p3]];
    }
    flip = dist1 ? dist1 < 0 : dist2 < 0;
  }
  return flip ? hull.reverse() : hull;
}
```

**Comment by iconexperience** [comment]
@lehni Sorry, I did not test with your example, I only saw that generally we were getting tDiffs around `-1e-16` in our test cycles. That one seems to be special and I will take a closer look.

**Comment by lehni** [comment]
Even more condensed, with comments : )

```js
    function getConvexHull(dq0, dq1, dq2, dq3) {
        var p0 = [ 0, dq0 ],
            p1 = [ 1 / 3, dq1 ],
            p2 = [ 2 / 3, dq2 ],
            p3 = [ 1, dq3 ],
            // Find signed distance of p1 and p2 from line [p0, p3]
            getSignedDistance = Line.getSignedDistance,
            dist1 = getSignedDistance(0, dq0, 1, dq3, 1 / 3, dq1),
            dist2 = getSignedDistance(0, dq0, 1, dq3, 2 / 3, dq2),
            hull;
        // Check if p1 and p2 are on the same side of the line [p0, p3]
        if (dist1 * dist2 < 0) {
            hull = [[p0, p1, p3], [p0, p2, p3]];
        } else {
            var distRatio = dist1 / dist2;
             hull = [
                // p2 is inside the triangle, hull is a triangle.
                distRatio >= 2 ? [p0, p1, p3]
                 // p1 is inside the triangle, hull is a triangle.
                : distRatio <= .5 ? [p0, p2, p3]
                // Hull is a quadrilateral, we need all lines in correct order.
                : [p0, p1, p2, p3],
                // Line [p0, p3] is part of the hull.
                [p0, p3]
            ];
        }
        return (dist1 || dist2) < 0 ? hull.reverse() : hull;
    }
```

**Comment by iconexperience** [comment]
@lehni Can you make a condensed version of my fixed implementation?

**Comment by lehni** [comment]
Yes I was just going to write that.

**Comment by lehni** [comment]
Here we go. Amazing work, @iconexperience. And the error is gone!

```js
    function getConvexHull(dq0, dq1, dq2, dq3) {
        var p0 = [ 0, dq0 ],
            p1 = [ 1 / 3, dq1 ],
            p2 = [ 2 / 3, dq2 ],
            p3 = [ 1, dq3 ],
            // Find signed distance of p1 and p2 from line [p0, p3]
            dist1 = dq1 - (dq0 + (dq3 - dq0) / 3),
            dist2 = dq2 - (dq0 + 2 * (dq3 - dq0) / 3),
            hull;
        // Check if p1 and p2 are on the opposite sides of the line [p0, p3]
        if (dist1 * dist2 < 0) {
            hull = [[p0, p1, p3], [p0, p2, p3]];
        } else {
            var distRatio = dist1 / dist2;
             hull = [
                // p2 is inside, the hull is a triangle.
                distRatio >= 2 ? [p0, p1, p3]
                // p1 is inside, the hull is a triangle.
                : distRatio <= .5 ? [p0, p2, p3]
                // Hull is a quadrilateral, we need all lines in correct order.
                : [p0, p1, p2, p3],
                // Line [p0, p3] is part of the hull.
                [p0, p3]
            ];
        }
        return (dist1 || dist2) < 0 ? hull.reverse() : hull;
    }
```

**Comment by iconexperience** [comment]
Of course it is gone :smile:

**Comment by lehni** [comment]
Very cool! Would you like to improve the comments a bit and port some over from the original implementation, so they make sense with your changes? I think you're in a better position to do so than me. You could create a PR for it, or I could simply cherry pick it from your branch.

**Comment by lehni** [comment]
I think the error that you describe above is also what was wrong in the original implementation, correct?

**Comment by iconexperience** [comment]
Yes, i will do that, but later today.

Oh, I think we do not need that ugly hack to improve precision of `getSignedDistance()` anymore. I did not really like that, so I suggest we try to stick with the old implementation.

**Comment by lehni** [comment]
All sounds good with me! I had a feeling we started a witch hunt there...

**Comment by iconexperience** [comment]
I think that the original implementation not really buggy, it was just unnecessarily complicated and maybe it ran into numeric precision problems. I really do not know why your example had such a strange `tDiff` with the old version.

**Comment by lehni** [comment]
BTW, I checked these latest code changes against the `boolean-test` suite, and it appears that reducing `epslion` to `1e-8` creates new issues there... `1e-7` appears to be fine thought, and I had that in place for a while earlier already. We could also test `CURVETIME_EPSILON / 2`. But the two additional gltiches that go away when lowering to `1e-8` might be caused by other things... We'll have to hand-inspect those cases anyway.

**Comment by iconexperience** [comment]
So I think we do not need another issue on fat line clipping. The changes that I would change are

1. The new `getConvexHull()` function.
2. Use 0.76 istead of 3/4 for factor. In one case this really makes a difference, and if 3/4 really was the upper limit, we should not see any changes when using a higher factor.
3. Small improvement to <a href="https://github.com/paperjs/paper.js/blob/boolean-fix/src/path/Curve.js#L1606">`clipConvexHullPart()`</a> to better handle cases when `qy == threshold`. This cannot have any negative side effects.

With these changes in place I am down to 21 error messages in the test kit, if I tested correctly.

**Comment by lehni** [comment]
Very cool. I agree, no additional issue is required. I merged all requests in. Regarding point 2:

I don't really understand the specifics of these factors... It seems a bit strange to pick such an arbitrary value though. How would 0.8 be more problematic? What about 3 / 4 + epsilon?

**Comment by lehni** [comment]
BTW: I am working on the linking and better interpretation over multiple intersections in the same spot, and have the glitches down to 18. There is some special handling for overlaps that I need to implement, I have a feeling it will solve many of the remaining issues.

**Comment by iconexperience** [comment]
Oh, you don't remember? Take a look at https://github.com/paperjs/paper.js/issues/765#issuecomment-141523770

That's the same factor.

**Comment by iconexperience** [comment]
Great to hear about the other progress. I will try to find the curve pair where 3/4 does not work and handle this is a new issue, so we have it documented. There is another case that used to take 20 seconds to solve when I removed the last condition in `addCurveIntersections()`. I want to understand what happened there. If you want me to do something else, let me know.

**Comment by lehni** [comment]
No that sounds good! Meanwhile, i got some ugly code running here that reduces the glitches to 10 : )

**Comment by iconexperience** [comment]
That sounds just incredible. Keep going, Jürg!

**Comment by lehni** [comment]
Check this out:
```
glitches: 10, 0 new, 8 solved
errors: 0, 0 new, 0 solved
boolCount: 42567
glitches = [163, 1428, 1920, 2765, 3214, 4693, 5321, 5419, 5615, 7787]
new glitches = []
solved glitches = [813, 1628, 1629, 3636, 5770, 7469, 7777, 7778]
errors = []
```
: )

**Comment by iconexperience** [comment]
Fantastic!

Now replace <a href="https://github.com/paperjs/paper.js/blob/boolean-fix/src/path/Curve.js#L1475">Curve.js lines 1475/1476</a>  with the following lines and let me know if that kills any of the remaining glitches.

```
tMinNew = tMin + (tMax - tMin) * tMinClip,
tMaxNew = tMin + (tMax - tMin) * tMaxClip;
```

I think it will!

**Comment by lehni** [comment]
I'm away from the computer for quite a while now but will sure try this out later!

**Comment by lehni** [comment]
I just tested this. It hasn't changed any of the remaining 10 glitches, but maybe its effects are overshadowed by my other changes that are now handling overlaps in a more robust way? I will commit this either way, since it is simpler.

**Comment by iconexperience** [comment]
Too bad, it actually reduced the number of glitches here by 2. I was quite surprised, but as you said, the code is simpler and easier to understand.

Regarding the 3/4 factor I am not so sure if I was right. Changing this to a slightly higher value affects all following recursions. This means the one solved glitch could be pure coincidence, therefore I decided to stick with the 3/4 for now.

**Comment by lehni** [comment]
Yes that's what I meant with the witch-hunt : )

**Comment by iconexperience** [comment]
I just now realized what you wanted to tell me <a href="https://github.com/paperjs/paper.js/issues/795#issuecomment-145673278">here</a>. Well, we are all a bit stubborn from time to time. At least I am :)

**Comment by lehni** [comment]
Haha, oh well I think stubbornness is also what's getting us so close towards the goal so quickly, so nothing wrong with that! I think it's great to keep questioning things that were thought to be long set in stone... I just had a hunch about this being the wrong place to look and tried to verbalize it : )

````
glitches: 5, 0 new, 5 solved
errors: 0, 0 new, 0 solved
boolCount: 42567
glitches = [163, 3214, 4693, 5615, 7787]
new glitches = []
solved glitches = [1428, 1920, 2765, 5321, 5419]
errors = []
new errors = []
solved errors = []
```

**Comment by iconexperience** [comment]
Oh, and I just realized that changing epsilon to `1e-8` or `1e-9`doesn't improve things anymore, so that's off the table. 

So let me summarize the solution of this issue. What we have done is:

1. Remove the convergence condition in `addCurveIntersections()`, which seemed to cause more harm than good, but we should keep an eye on that.
2. Implement a better version of `getConvexCurve()` that also solved the issue with negative `tDiff`values.

Also I proposed a lot of changes that turned out to be not so great ideas.

That's basically it.
